integreat_cms/cms/models/contact/contact.py

```diff
@@ -54,6 +62,39 @@ def region(self) -> Region:
         """
         return self.location.region
 
+    @classmethod
+    def search(cls, region: Region, query: str) -> QuerySet:
+        """
+        Searches for all contacts which match the given `query` in their comment.
+        :param region: The current region
+        :param query: The query string used for filtering the contacts
+        :return: A query for all matching objects
+        """
+        searchable_fields = ("title", "name", "email", "phone_number", "website")
+
+        q = models.Q()
+
+        for word in query.split():
+            # Every word has to appear in at least one field
+            OR = [
+                models.Q(**{f"{field}__icontains": word}) for field in searchable_fields
+            ]
+            # We OR whether it appears in each of the field, and
+            # AND those expressions corresponding to each word
+            # because we are not interested in objects where one word is missing
+            q &= reduce(lambda a, b: a | b, OR)
+
+        # We could add annotations to determine how closely each result matches the query,
+        # e.g. by finding the length of the longest common substring between each field and the original query,
+        # taking the square of that value to obtain something representing the "contribution" of that field
+        # (so longer matches in only a few fields get a much higher value than many short matches all over)
+        # and then summing those together to obtain an overall score of how relevant that object is to the query,
+        # but that would require us find the longest common substring on the db level,
+        # and that feels a bit overkill for now  (it will likely be confusing to re-discover and maintain,
+        # especially if we were to also employ fuzzy matching – which would be much preferred, if we can do it)
+
+        return cls.objects.filter(q, location__region=region)
+
```

Comment:
This is definitely a respectable approach. I think implementing fuzzy search ourself is overkill, and not really maintainable in the long run.

Did you loook into exitseng solutions for this? E.g.:
https://django-haystack.readthedocs.io/en/master/toc.html
https://gist.github.com/shahriar350/aebcaef26c79c575834f4c89058cf6f3

Or (probably a lot easier): https://docs.djangoproject.com/en/5.1/ref/contrib/postgres/search/#searchvector

Reply:
To be honest, barely, and I stopped at trigram similarity. I did a brief search I think, but I might have already been too deep in the technical details. I do tend to fail to look beyond the documentation and capabilities of the system right in front of me for dependencies to add that might solve this